lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -1468,6 +1476,78 @@ Value FIRRTLLowering::getLoweredValue(Value value) {
+/// We have to care about the extension/truncation/signedness of each element.
+/// If returns a null value for complex arrays such as arrays with bundles.
+Value FIRRTLLowering::getExtOrTruncArrayValue(Value array,
+                                              FIRRTLType sourceType,
+                                              FIRRTLType destType,
+                                              bool allowTruncate) {
+  SmallVector<Value> resultBuffer;
+  bool success = true;
+
+  // Helper function to cast each element of array to dest type.
+  auto cast = [&](Value value, FIRRTLType sourceType, FIRRTLType destType) {
+    auto srcWidth = sourceType.cast<IntType>().getWidthOrSentinel();
+    auto destWidth = destType.cast<IntType>().getWidthOrSentinel();
+    auto resultType = builder.getIntegerType(destWidth);
+
+    if (srcWidth == destWidth)
+      return value;
+
+    if (srcWidth > destWidth) {
+      if (allowTruncate)
+        return builder.createOrFold<comb::ExtractOp>(resultType, value, 0);
+
+      builder.emitError("operand should not be a truncation");
+      success = false;
+      return Value();
+    }
+
+    if (sourceType.cast<IntType>().isSigned())
+      return builder.createOrFold<comb::SExtOp>(resultType, value);
+    auto zero = getOrCreateIntConstant(destWidth - srcWidth, 0);
+    return builder.createOrFold<comb::ConcatOp>(zero, value);
+  };
+
+  // This recursive function constructs the output array.
+  std::function<void(Value, FIRRTLType, FIRRTLType)> recurse =
+      [&](Value src, FIRRTLType srcType, FIRRTLType destType) {
+        TypeSwitch<FIRRTLType>(srcType)
+            .Case<FVectorType>([&](auto srcVectorType) {
+              auto destVectorType = destType.cast<FVectorType>();
+              int size = resultBuffer.size();
+              for (size_t i = 0, e = std::min(srcVectorType.getNumElements(),
+                                              destVectorType.getNumElements());
+                   i != e; ++i) {
+                auto iIdx = getOrCreateIntConstant(
+                    llvm::Log2_64_Ceil(srcVectorType.getNumElements()), i);
+                auto arrayIndex = builder.create<hw::ArrayGetOp>(src, iIdx);
+                recurse(arrayIndex, srcVectorType.getElementType(),
+                        destVectorType.getElementType());
+                if (!success)
+                  return;
+              }
+              SmallVector<Value> temp(resultBuffer.begin() + size,
+                                      resultBuffer.end());
+              auto array = builder.createOrFold<hw::ArrayCreateOp>(temp);
+              resultBuffer.resize(size);
+              resultBuffer.push_back(array);
+            })
+            .Case<IntType>([&](auto) {
+              resultBuffer.push_back(cast(src, srcType, destType));
+            })
+            .Default([&](auto) { success = false; });
+      };
+
+  if (!success)
+    return Value();
```

Comment:
This looks like dead code.

Reply:
Yes, it was mistake. I changed to use `LogicalResult` for better code.